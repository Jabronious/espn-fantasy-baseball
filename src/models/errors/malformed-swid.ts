export class MalformedSWID extends Error {
	constructor(
		message = 'SWID does not match format: /^{[0-9A-F]{8}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{12}}$/'
	) {
		super(message);
		this.name = 'MalformedSWID';
	}
}
