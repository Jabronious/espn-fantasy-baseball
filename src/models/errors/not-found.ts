export class NotFound extends Error {
	constructor(message = 'Not Found') {
		super(message);
		this.name = 'NotFound';
	}
}
