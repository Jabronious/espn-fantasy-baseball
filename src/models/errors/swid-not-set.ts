export class SWIDNotSet extends Error {
	constructor(message = 'SWID has not been set') {
		super(message);
		this.name = 'SWIDNotSet';
	}
}
