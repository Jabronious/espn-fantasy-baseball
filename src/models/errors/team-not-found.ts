export class TeamNotFound extends Error {
	constructor(message = 'Team was not found in the league') {
		super(message);
		this.name = 'TeamNotFound';
	}
}
