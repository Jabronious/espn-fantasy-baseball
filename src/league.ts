import { BaseClass } from './base-class';
import { ESPNCookiesDto, SWID } from './models/classes/espn-cookies.dto';
import { LeagueDto } from './models/classes/league.dto';
import { MatchUpDto, MatchUpResponseDto } from './models/classes/matchup.dto';
import { MemberDto } from './models/classes/member.dto';
import { TeamDto } from './models/classes/team.dto';
import { Positions } from './models/enums/positions';
import { HittingStats, PitchingStats } from './models/enums/stats';
import { NotFound } from './models/errors/not-found';
import { Players } from './players';
import { Teams } from './teams';

type PositionCounts = { [key in Positions]?: number };
type HittingScoreStats = { [key in HittingStats]?: number };
type PitchingScoreStats = { [key in PitchingStats]?: number };

export class League extends BaseClass {
	positionCounts: PositionCounts = {};
	hittingScoreStats: HittingScoreStats = {};
	pitchingScoreStats: PitchingScoreStats = {};
	tradeDeadline?: number;
	constructor(leagueId: number, cookies?: ESPNCookiesDto) {
		super(leagueId, cookies);
	}

	/**
	 * @returns League - your currently initiated league
	 */
	async init(): Promise<League> {
		const params = {
			view: 'mSettings',
		};
		const leagueData = await this.fantasyRequests.get<LeagueDto>('', {}, params).catch((e: string) => {
			throw new Error(e);
		});

		const lineupSlotCounts = leagueData.settings.rosterSettings.lineupSlotCounts;
		Object.keys(leagueData.settings.rosterSettings.lineupSlotCounts).forEach((key) => {
			const positionKey = Positions[key as keyof typeof Positions];
			if (positionKey !== undefined) {
				this.positionCounts[positionKey] = lineupSlotCounts[key];
			}
		});

		leagueData.settings.scoringSettings.scoringItems.forEach((item: { statId: number; points: number }) => {
			if (HittingStats[item.statId]) {
				const statKey = HittingStats[item.statId as unknown as keyof typeof HittingStats];
				this.hittingScoreStats[statKey] = item.points;
			} else {
				const statKey = PitchingStats[item.statId as unknown as keyof typeof PitchingStats];
				this.pitchingScoreStats[statKey] = item.points;
			}
		});

		this.tradeDeadline = leagueData.settings.tradeSettings.deadlineDate;

		return this;
	}

	/**
	 * @returns Teams object
	 */
	teams(): Teams {
		return new Teams(this.leagueId, this.cookies);
	}

	players(): Players {
		return new Players(this.leagueId, this.cookies);
	}

	/**
	 * @returns a list of all team data
	 */
	async getLeagueTeams(): Promise<TeamDto[]> {
		const leagueData = await this.leagueRequest();
		return leagueData.teams;
	}

	/**
	 * @returns a list of members in the league
	 */
	async getLeagueMembers(): Promise<MemberDto[]> {
		const leagueData = await this.leagueRequest();
		return leagueData.members;
	}

	/**
	 * @returns a boolean whether or not the current SWID is league manager
	 */
	async isLeagueManager(): Promise<boolean> {
		const members = await this.getLeagueMembers();
		const swid = new SWID(this.cookies?.swid).swid;

		if (members === undefined) {
			throw new NotFound('No Members Found');
		}

		const member = members.find((member: MemberDto) => member.id === swid);

		if (!member) return false;

		return member.isLeagueManager;
	}

	/**
	 * @returns a list of matches for the given week
	 */
	async getWeeklyMatchups(): Promise<MatchUpDto[]> {
		const params = {
			view: 'mMatchup',
		};
		const matchUpsRequest = await this.fantasyRequests.get<MatchUpResponseDto>('', {}, params).catch((e) => {
			throw new Error(e);
		});

		const currentMatchUps = matchUpsRequest.schedule.filter(
			(matchUp: MatchUpDto) => matchUp.matchupPeriodId === matchUpsRequest.gameId
		);

		return currentMatchUps;
	}

	private async leagueRequest(): Promise<LeagueDto> {
		return await this.fantasyRequests.get<LeagueDto>().catch((e) => {
			throw new Error(e);
		});
	}
}
