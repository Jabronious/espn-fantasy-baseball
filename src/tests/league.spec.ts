import axios from 'axios';
import { League } from '../league';
import { ESPNCookiesDto } from '../models/classes/espn-cookies.dto';
import { NotFound } from '../models/errors/not-found';

describe('League', () => {
	const cookies: ESPNCookiesDto = {
		espn_s2: '2B7uOaY7pTx8neDj061z24T3PdDPn6a29EHEMEfelzI6HLT43oy1P%%2BUTkzYAsk%2BkxgrK%',
		swid: '{A11A1111-1AAA-111A-1A11-111AAA111A1A}',
	};
	const league = new League(1, cookies);

	it('should be defined', () => {
		expect(league).toBeDefined();
	});

	describe('init', () => {
		it('should get init team data', async () => {
			jest.spyOn(axios, 'request').mockResolvedValueOnce({
				data: {
					settings: {
						rosterSettings: {
							lineupSlotCounts: {
								'0': 1,
								'1': 1,
								'2': 1,
								'3': 1,
								'4': 1,
								'5': 5,
								'6': 1,
								'7': 1,
								'8': 0,
								'9': 0,
								'10': 0,
								'11': 0,
								'12': 1,
								'13': 0,
								'14': 5,
								'15': 3,
								'16': 4,
								'17': 3,
								'18': 0,
								'19': 0,
							},
						},
						scoringSettings: {
							scoringItems: [
								{
									isReverseItem: false,
									leagueRanking: 0.0,
									leagueTotal: 0.0,
									points: 1.0,
									statId: 20,
								},
								{
									isReverseItem: false,
									leagueRanking: 0.0,
									leagueTotal: 0.0,
									points: 1.0,
									statId: 76,
								},
							],
						},
						tradeSettings: {
							deadlineDate: 123412341234124,
						},
					},
				},
			});

			expect(await league.init()).toMatchSnapshot();
		});

		it('should catch errors and throw by init', async () => {
			jest.spyOn(axios, 'request').mockRejectedValueOnce('error');

			await expect(league.init()).rejects.toMatchSnapshot();
		});
	});

	describe('teams', () => {
		it('should be defined', () => {
			expect(league.teams()).toBeDefined();
		});
	});

	describe('players', () => {
		it('should be defined', () => {
			expect(league.players()).toBeDefined();
		});
	});

	describe('getLeagueTeams', () => {
		it('should get all league teams', async () => {
			jest.spyOn(axios, 'request').mockResolvedValueOnce({ data: { teams: { abbrev: 'CDMD' } } });

			expect(await league.getLeagueTeams()).toMatchSnapshot();
		});

		it('should catch errors and throw them', async () => {
			jest.spyOn(axios, 'request').mockRejectedValueOnce('error');

			await expect(league.getLeagueTeams()).rejects.toMatchSnapshot();
		});
	});

	describe('getLeagueMembers', () => {
		it('should get all league members', async () => {
			jest.spyOn(axios, 'request').mockResolvedValueOnce({
				data: { members: [{ displayName: 'string', id: 12345, isLeagueManager: true }] },
			});

			expect(await league.getLeagueMembers()).toMatchSnapshot();
		});

		it('should catch errors and throw them', async () => {
			jest.spyOn(axios, 'request').mockRejectedValueOnce('error');

			await expect(league.getLeagueMembers()).rejects.toMatchSnapshot();
		});
	});

	describe('isLeagueManager', () => {
		it('should return true when isLeagueManager is true', async () => {
			jest.spyOn(axios, 'request').mockResolvedValueOnce({
				data: { members: [{ displayName: 'string', id: cookies.swid, isLeagueManager: true }] },
			});

			expect(await league.isLeagueManager()).toBeTruthy();
		});

		it('should return false when isLeagueManager is false', async () => {
			jest.spyOn(axios, 'request').mockResolvedValueOnce({
				data: { members: [{ displayName: 'string', id: 'swid', isLeagueManager: false }] },
			});

			expect(await league.isLeagueManager()).toBeFalsy();
		});

		it('should return false when no members are found', async () => {
			jest.spyOn(axios, 'request').mockResolvedValueOnce({
				data: {},
			});

			await expect(league.isLeagueManager()).rejects.toThrowError(NotFound);
		});
	});

	describe('getWeeklyMatchups', () => {
		it('should get the weekly matchups', async () => {
			jest.spyOn(axios, 'request').mockResolvedValueOnce({
				data: { gameId: 1, schedule: [{ matchupPeriodId: 1 }] },
			});

			expect(await league.getWeeklyMatchups()).toMatchSnapshot();
		});

		it('should catch errors and throw them', async () => {
			jest.spyOn(axios, 'request').mockRejectedValueOnce('error');

			await expect(league.getWeeklyMatchups()).rejects.toMatchSnapshot();
		});
	});
});
