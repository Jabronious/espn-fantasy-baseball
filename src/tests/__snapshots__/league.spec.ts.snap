// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`League getLeagueMembers should catch errors and throw them 1`] = `[Error: error]`;

exports[`League getLeagueMembers should get all league members 1`] = `
Array [
  Object {
    "displayName": "string",
    "id": 12345,
    "isLeagueManager": true,
  },
]
`;

exports[`League getLeagueTeams should catch errors and throw them 1`] = `[Error: error]`;

exports[`League getLeagueTeams should get all league teams 1`] = `
Object {
  "abbrev": "CDMD",
}
`;

exports[`League getWeeklyMatchups should catch errors and throw them 1`] = `[Error: error]`;

exports[`League getWeeklyMatchups should get the weekly matchups 1`] = `
Array [
  Object {
    "matchupPeriodId": 1,
  },
]
`;

exports[`League init should catch errors and throw by init 1`] = `[Error: error]`;

exports[`League init should get init team data 1`] = `
League {
  "cookies": Object {
    "espn_s2": "2B7uOaY7pTx8neDj061z24T3PdDPn6a29EHEMEfelzI6HLT43oy1P%%2BUTkzYAsk%2BkxgrK%",
    "swid": "{A11A1111-1AAA-111A-1A11-111AAA111A1A}",
  },
  "fantasyRequests": FantasyRequest {
    "cookies": Object {
      "espn_s2": "2B7uOaY7pTx8neDj061z24T3PdDPn6a29EHEMEfelzI6HLT43oy1P%%2BUTkzYAsk%2BkxgrK%",
      "swid": "{A11A1111-1AAA-111A-1A11-111AAA111A1A}",
    },
    "endpoint": "https://lm-api-reads.fantasy.espn.com/apis/v3/games/flb/seasons/2024/segments/0/leagues/1",
  },
  "hittingScoreStats": Object {
    "R": 1,
  },
  "leagueId": 1,
  "pitchingScoreStats": Object {
    "PTW": 1,
  },
  "positionCounts": Object {
    "1B": 1,
    "1B/3B": 1,
    "2B": 1,
    "2B/SS": 1,
    "3B": 1,
    "??": 0,
    "BE": 4,
    "C": 1,
    "CF": 0,
    "DH": 0,
    "IF": 0,
    "IL": 3,
    "LF": 0,
    "OF": 5,
    "P": 0,
    "RF": 0,
    "RP": 3,
    "SP": 5,
    "SS": 1,
    "UTIL": 1,
  },
  "tradeDeadline": 123412341234124,
}
`;
